# SQL CARTO QUERY FUNCTION
# written for PUI2016_Python2

##### This notebook is set up to link as a default to the fb55 account.  To turn in the homework use the same account you used in the lab (hvt201)  and query the database that you were querying in class (citibike_feb_2015)

In [1]:
SQL_SOURCE = 'https://fb55.carto.com/api/v2/sql?q='

import urllib2
import urllib
import StringIO
import ast
import pandas as pd

def queryCartoDB(query, format='CSV', source=SQL_SOURCE):
    '''queries carto datasets from a given carto account
    Arguments: 
    query - string: a valid sql query string
    format - outlut format  OPTIONAL (default CSV)
    source - a valid sql api endpoint OPTIONAL (default carto fb55 account)
    Returns:
    the return of the sql query AS A STRING
    NOTES:
    designed for the carto API, tested only with CSV return format'''
    
    data = urllib.urlencode({'format': format, 'q': query})
    try:
        response = urllib2.urlopen(source, data)
    except urllib2.HTTPError,e:
        raise ValueError('\n'.join(ast.literal_eval(e.readline())['error']))
    except Exception:
        raise
    return response.read()

In [2]:
test = '''
SELECT *
FROM citibike
'''
pd.read_csv(StringIO.StringIO(queryCartoDB(test)), sep=',')

the_geom        start_station_name  the_geom_webmercator  field_1  \
0           NaN        E 20 St & Park Ave                   NaN      175   
1           NaN          W 43 St & 10 Ave                   NaN     1159   
2           NaN         E 6 St & Avenue B                   NaN     2827   
3           NaN           E 12 St & 3 Ave                   NaN     4961   
4           NaN           W 41 St & 8 Ave                   NaN     6156   
5           NaN           5 Ave & E 29 St                   NaN     7433   
6           NaN      Allen St & Hester St                   NaN     7779   
7           NaN           E 39 St & 3 Ave                   NaN     8327   
8           NaN          10 Ave & W 28 St                   NaN     9435   
9           NaN            E 2 St & 2 Ave                   NaN    10472   
10          NaN           1 Ave & E 44 St                   NaN    10931   
11          NaN           W 45 St & 6 Ave                   NaN    11361   
12          NaN           W 18 St & 6 Ave                   NaN    12173   
13          NaN           E 58 St & 3 Ave                   NaN    12211   
14          NaN           E 33 St & 5 Ave                   NaN    12560   
15          NaN           W 31 St & 7 Ave                   NaN    13239   
16          NaN           5 Ave & E 29 St                   NaN    13365   
17          NaN           E 39 St & 2 Ave                   NaN    13750   
18          NaN          10 Ave & W 28 St                   NaN    14218   
19          NaN           W 42 St & 8 Ave                   NaN    14431   
20          NaN           E 27 St & 1 Ave                   NaN    15157   
21          NaN            Great Jones St                   NaN    15669   
22          NaN           W 49 St & 8 Ave                   NaN    16526   
23          NaN        Broadway & W 24 St                   NaN    16572   
24          NaN           E 40 St & 5 Ave                   NaN    18931   
25          NaN           E 16 St & 5 Ave                   NaN    19229   
26          NaN        Broadway & W 36 St                   NaN    20481   
27          NaN           8 Ave & W 31 St                   NaN    21285   
28          NaN           W 45 St & 6 Ave                   NaN    23338   
29          NaN       FDR Drive & E 35 St                   NaN    23392   
...         ...                       ...                   ...      ...   
46170       NaN  Washington Pl & Broadway                   NaN    46170   
46171       NaN      Allen St & Hester St                   NaN    46171   
46172       NaN     Warren St & Church St                   NaN    46172   
46173       NaN           E 12 St & 3 Ave                   NaN    46173   
46174       NaN           Dean St & 4 Ave                   NaN    46174   
46175       NaN      Allen St & Hester St                   NaN    46175   
46176       NaN        Broadway & E 22 St                   NaN    46176   
46177       NaN     Washington Pl & 6 Ave                   NaN    46177   
46178       NaN           E 31 St & 3 Ave                   NaN    46178   
46179       NaN        Broadway & W 60 St                   NaN    46179   
46180       NaN        Carmine St & 6 Ave                   NaN    46180   
46181       NaN           W 52 St & 5 Ave                   NaN    46181   
46182       NaN           Dean St & 4 Ave                   NaN    46182   
46183       NaN           9 Ave & W 45 St                   NaN    46183   
46184       NaN        Broadway & W 58 St                   NaN    46184   
46185       NaN        Broadway & W 58 St                   NaN    46185   
46186       NaN   Watts St & Greenwich St                   NaN    46186   
46187       NaN           W 51 St & 6 Ave                   NaN    46187   
46188       NaN   University Pl & E 14 St                   NaN    46188   
46189       NaN           W 11 St & 6 Ave                   NaN    46189   
46190       NaN           W 17 St & 8 Ave              

### Task 1 — Familiarize with SQL Clauses

#### 1. Sort data by start_station_id, tripduration
     - Only checking trips with duration <= 3 hours
#### 3. Only show the top/last 10 records (aka head and tail in SQL) List all unique start_station_id values
#### 4. Aggregation functions:
#### 5. Count the number of trips (aka wc -l in SQL) Find the average/min/max trip duration

In [3]:
author = "__Gabriel__"
contact = "__gabrielyin@berkeley.edu__"

"""
Copy at your own risk!! If you have questions I am willing to help. Just don't copy and paste my code. 
You won't learn anything from it. I will report, if necessary, suspicious plagiarism to Student Judicial Affairs.

"""

# begin task1 
# first 10 rows 
task1_1 = '''
select * from citibike
where tripduration <= 60 * 60 * 3
order by start_station_id
limit 10
'''
pd.read_csv(StringIO.StringIO(queryCartoDB(task1_1)), sep=',')

the_geom start_station_name  the_geom_webmercator  field_1  \
0       NaN   W 52 St & 11 Ave                   NaN     3370   
1       NaN   W 52 St & 11 Ave                   NaN     4024   
2       NaN   W 52 St & 11 Ave                   NaN     2348   
3       NaN   W 52 St & 11 Ave                   NaN     2051   
4       NaN   W 52 St & 11 Ave                   NaN      455   
5       NaN   W 52 St & 11 Ave                   NaN     1597   
6       NaN   W 52 St & 11 Ave                   NaN     2208   
7       NaN   W 52 St & 11 Ave                   NaN      718   
8       NaN   W 52 St & 11 Ave                   NaN     1299   
9       NaN   W 52 St & 11 Ave                   NaN     5202   

   start_station_longitude  tripduration               starttime  \
0               -73.993929           569  2015-02-01 14:52:00+00   
1               -73.993929           387  2015-02-01 15:57:00+00   
2               -73.993929           301  2015-02-01 13:08:00+00   
3               -73.993929          1944  2015-02-01 12:39:00+00   
4               -73.993929           961  2015-02-01 08:38:00+00   
5               -73.993929           962  2015-02-01 11:47:00+00   
6               -73.993929           339  2015-02-01 12:54:00+00   
7               -73.993929           344  2015-02-01 09:49:00+00   
8               -73.993929           934  2015-02-01 11:14:00+00   
9               -73.993929           455  2015-02-01 17:55:00+00   

                 stoptime  start_station_id  cartodb_id  \
0  2015-02-01 15:01:00+00                72        3302   
1  2015-02-01 16:03:00+00                72        3957   
2  2015-02-01 13:13:00+00                72        2282   
3  2015-02-01 13:12:00+00                72        1983   
4  2015-02-01 08:54:00+00                72         387   
5  2015-02-01 12:03:00+00                72        1527   
6  2015-02-01 13:00:00+00                72        2142   
7  2015-02-01 09:54:00+00                72         649   
8  2015-02-01 11:30:00+00                72        1230   
9  2015-02-01 18:03:00+00                72        5136   

   start_station_latitude  end_station_id         end_station_name  \
0               40.767272             499       Broadway & W 60 St   
1               40.767272             514         12 Ave & W 40 St   
2               40.767272             447          8 Ave & W 52 St   
3               40.767272             328  Watts St & Greenwich St   
4               40.767272             212  W 16 St & The High Line   
5               40.767272             212  W 16 St & The High Line   
6               40.767272             478         11 Ave & W 41 St   
7               40.767272            2021          W 45 St & 8 Ave   
8               40.767272             212  W 16 St & The High Line   
9               40.767272             529          W 42 St & 8 Ave   

   end_station_latitude  end_station_longitude  bikeid    usertype  \
0             40.769155             -73.981918   17429  Subscriber   
1             40.760875             -74.002777   15282  Subscriber   
2             40.763707             -73.985162   18174  Subscriber   
3             40.724055             -74.009660   17849    Customer   
4             40.743349             -74.006818   20786  Subscriber   
5             40.743349             -74.006818   16524  Subscriber   
6             40.760301             -73.998842   19672  Subscriber   
7             40.759291             -73.988597   14692  Subscriber   
8             40.743349             -74.006818   17541  Subscriber   
9             40.757570             -73.990985   16427  Subscriber   

   birth_year  gender  
0      1981.0       1  
1      1991.0       1  
2      1974.0       1  
3         NaN       0  
4      1972.0       2  
5      1977.0       1  
6      1982.0       1  
7      1986.0       2  
8      1969.0       1  
9      1983.0       1

In [4]:
# last 10 rows 
task1_2 = '''
select * from citibike
where tripduration <= 60 * 60 * 3
order by start_station_id desc
limit 10
'''
pd.read_csv(StringIO.StringIO(queryCartoDB(task1_2)), sep=',')

the_geom          start_station_name  the_geom_webmercator  field_1  \
0       NaN  South End Ave & Liberty St                   NaN     1932   
1       NaN  South End Ave & Liberty St                   NaN     2094   
2       NaN  South End Ave & Liberty St                   NaN     1698   
3       NaN  South End Ave & Liberty St                   NaN     1657   
4       NaN  South End Ave & Liberty St                   NaN     1402   
5       NaN  South End Ave & Liberty St                   NaN     1542   
6       NaN  South End Ave & Liberty St                   NaN      361   
7       NaN  South End Ave & Liberty St                   NaN     1297   
8       NaN  South End Ave & Liberty St                   NaN     1576   
9       NaN  South End Ave & Liberty St                   NaN     2271   

   start_station_longitude  tripduration               starttime  \
0               -74.015756           689  2015-02-01 12:25:00+00   
1               -74.015756           307  2015-02-01 12:42:00+00   
2               -74.015756           745  2015-02-01 11:57:00+00   
3               -74.015756           858  2015-02-01 11:53:00+00   
4               -74.015756          1243  2015-02-01 11:27:00+00   
5               -74.015756           683  2015-02-01 11:43:00+00   
6               -74.015756           301  2015-02-01 07:49:00+00   
7               -74.015756          1900  2015-02-01 11:14:00+00   
8               -74.015756          7225  2015-02-01 11:46:00+00   
9               -74.015756           790  2015-02-01 13:01:00+00   

                 stoptime  start_station_id  cartodb_id  \
0  2015-02-01 12:36:00+00              3002        1863   
1  2015-02-01 12:47:00+00              3002        2026   
2  2015-02-01 12:09:00+00              3002        1628   
3  2015-02-01 12:07:00+00              3002        1587   
4  2015-02-01 11:48:00+00              3002        1333   
5  2015-02-01 11:55:00+00              3002        1472   
6  2015-02-01 07:54:00+00              3002         293   
7  2015-02-01 11:45:00+00              3002        1228   
8  2015-02-01 13:47:00+00              3002        1506   
9  2015-02-01 13:14:00+00              3002        2205   

   start_station_latitude  end_station_id               end_station_name  \
0               40.711512             358  Christopher St & Greenwich St   
1               40.711512             147       Greenwich St & Warren St   
2               40.711512             369          Washington Pl & 6 Ave   
3               40.711512             355          Bayard St & Baxter St   
4               40.711512             360           William St & Pine St   
5               40.711512             308        St James Pl & Oliver St   
6               40.711512             417         Barclay St & Church St   
7               40.711512             525               W 34 St & 11 Ave   
8               40.711512             499             Broadway & W 60 St   
9               40.711512             368             Carmine St & 6 Ave   

   end_station_latitude  end_station_longitude  bikeid    usertype  \
0             40.732916             -74.007114   19309  Subscriber   
1             40.715422             -74.011220   18386  Subscriber   
2             40.732241             -74.000264   20901  Subscriber   
3             40.716021             -73.999744   15748  Subscriber   
4             40.707179             -74.008873   18288    Customer   
5             40.713079             -73.998512   16987  Subscriber   
6             40.712912             -74.010202   19484  Subscriber   
7             40.755942             -74.002116   17849    Customer   
8             40.769155             -73.981918   21202    Customer   
9             40.730386             -74.002150   17004  Subscriber   

   birth_year  gender  
0      1996.0       1  
1      1984.0       2  
2      1980.0       1  
3      1959.0       1  
4         NaN       0  
5      1980.0       2  
6      1980.0

In [5]:
# sort by tripduration: first 10 rows 
task1_3 = '''
select * from citibike
where tripduration <= 60 * 60 * 3
order by tripduration
limit 10
'''
# sort by tripduration: last 10 rows 
task1_4 = '''
select * from citibike
where tripduration <= 60 * 60 * 3
order by tripduration desc
limit 10 
'''
pd.read_csv(StringIO.StringIO(queryCartoDB(task1_3)), sep=',')

the_geom       start_station_name  the_geom_webmercator  field_1  \
0       NaN          W 29 St & 9 Ave                   NaN    13533   
1       NaN    Laight St & Hudson St                   NaN    19295   
2       NaN      Mott St & Prince St                   NaN     8347   
3       NaN          8 Ave & W 31 St                   NaN    13098   
4       NaN          W 31 St & 7 Ave                   NaN     7865   
5       NaN     Division St & Bowery                   NaN    14163   
6       NaN        E 6 St & Avenue B                   NaN     3178   
7       NaN          5 Ave & E 29 St                   NaN     7364   
8       NaN  Grand St & Havemeyer St                   NaN    11165   
9       NaN          W 38 St & 8 Ave                   NaN    21463   

   start_station_longitude  tripduration               starttime  \
0               -73.998393            60  2015-02-04 07:52:00+00   
1               -74.007718            60  2015-02-04 18:22:00+00   
2               -73.994800            60  2015-02-03 08:07:00+00   
3               -73.994811            60  2015-02-04 06:54:00+00   
4               -73.991600            60  2015-02-02 21:39:00+00   
5               -73.997047            60  2015-02-04 08:41:00+00   
6               -73.981854            60  2015-02-01 14:34:00+00   
7               -73.986831            60  2015-02-02 17:14:00+00   
8               -73.956981            60  2015-02-03 18:01:00+00   
9               -73.991382            60  2015-02-04 22:38:00+00   

                 stoptime  start_station_id  cartodb_id  \
0  2015-02-04 07:53:00+00               512       13482   
1  2015-02-04 18:23:00+00               248       19243   
2  2015-02-03 08:08:00+00               251        8284   
3  2015-02-04 06:55:00+00               521       13043   
4  2015-02-02 21:40:00+00               379        7798   
5  2015-02-04 08:42:00+00               296       14117   
6  2015-02-01 14:35:00+00               317        3110   
7  2015-02-02 17:15:00+00               474        7299   
8  2015-02-03 18:02:00+00               471       11104   
9  2015-02-04 22:39:00+00               523       21407   

   start_station_latitude  end_station_id         end_station_name  \
0               40.750073             512          W 29 St & 9 Ave   
1               40.721854             248    Laight St & Hudson St   
2               40.723180             251      Mott St & Prince St   
3               40.750450             521          8 Ave & W 31 St   
4               40.749156             379          W 31 St & 7 Ave   
5               40.714131             296     Division St & Bowery   
6               40.724537             317        E 6 St & Avenue B   
7               40.745168             474          5 Ave & E 29 St   
8               40.712868             471  Grand St & Havemeyer St   
9               40.754666             523          W 38 St & 8 Ave   

   end_station_latitude  end_station_longitude  bikeid    usertype  \
0             40.750073             -73.998393   21627  Subscriber   
1             40.721854             -74.007718   20309  Subscriber   
2             40.723180             -73.994800   19896  Subscriber   
3             40.750450             -73.994811   14788  Subscriber   
4             40.749156             -73.991600   17643  Subscriber   
5             40.714131             -73.997047   21644  Subscriber   
6             40.724537             -73.981854   20572  Subscriber   
7             40.745168             -73.986831   20842  Subscriber   
8             40.712868             -73.956981   16340  Subscriber   
9             40.754666             -73.991382   16148  Subscriber   

   birth_year  gender  
0        1960       2  
1        1970       1  
2        1982       1  
3        1971       1  
4        1986       1  
5        1990       1  
6        1985       1  
7        1956       1  
8        1980       1  
9        1980       1

In [6]:
pd.read_csv(StringIO.StringIO(queryCartoDB(task1_4)), sep=',')

the_geom             start_station_name  the_geom_webmercator  field_1  \
0       NaN          Greenwich Ave & 8 Ave                   NaN    41403   
1       NaN                W 29 St & 9 Ave                   NaN    10975   
2       NaN  Washington St & Gansevoort St                   NaN    40237   
3       NaN           Centre St & Worth St                   NaN     7538   
4       NaN             Broadway & W 55 St                   NaN     9756   
5       NaN                W 21 St & 6 Ave                   NaN    15380   
6       NaN             Broadway & W 58 St                   NaN    25241   
7       NaN             Broadway & E 22 St                   NaN    19952   
8       NaN             Broadway & W 60 St                   NaN     7139   
9       NaN        DeKalb Ave & Hudson Ave                   NaN    11871   

   start_station_longitude  tripduration               starttime  \
0               -74.002638         10757  2015-02-07 12:17:00+00   
1               -73.998393         10660  2015-02-03 17:42:00+00   
2               -74.008119         10174  2015-02-07 09:31:00+00   
3               -74.002345          9969  2015-02-02 18:22:00+00   
4               -73.981923          9938  2015-02-03 14:10:00+00   
5               -73.994156          9807  2015-02-04 10:53:00+00   
6               -73.981693          9711  2015-02-05 11:12:00+00   
7               -73.989551          9572  2015-02-04 19:09:00+00   
8               -73.981918          9556  2015-02-02 13:25:00+00   
9               -73.981013          9373  2015-02-03 19:25:00+00   

                 stoptime  start_station_id  cartodb_id  \
0  2015-02-07 15:16:00+00               284       41392   
1  2015-02-03 20:40:00+00               512       10912   
2  2015-02-07 12:20:00+00               405       40221   
3  2015-02-02 21:08:00+00               386        7472   
4  2015-02-03 16:55:00+00               468        9696   
5  2015-02-04 13:36:00+00               435       15336   
6  2015-02-05 13:54:00+00               457       25203   
7  2015-02-04 21:48:00+00               402       19897   
8  2015-02-02 16:05:00+00               499        7073   
9  2015-02-03 22:01:00+00               324       11810   

   start_station_latitude  end_station_id               end_station_name  \
0               40.739017             173             Broadway & W 49 St   
1               40.750073             435                W 21 St & 6 Ave   
2               40.739323             405  Washington St & Gansevoort St   
3               40.714948             264           Maiden Ln & Pearl St   
4               40.765265            2023        E 55 St & Lexington Ave   
5               40.741740             435                W 21 St & 6 Ave   
6               40.766953            2021                W 45 St & 8 Ave   
7               40.740343             517          Pershing Square South   
8               40.769155             499             Broadway & W 60 St   
9               40.689888             324        DeKalb Ave & Hudson Ave   

   end_station_latitude  end_station_longitude  bikeid    usertype  \
0             40.760647             -73.984427   15082  Subscriber   
1             40.741740             -73.994156   16099  Subscriber   
2             40.739323             -74.008119   21323  Subscriber   
3             40.707065             -74.007319   19921  Subscriber   
4             40.759681             -73.970314   19070  Subscriber   
5             40.741740             -73.994156   18088  Subscriber   
6             40.759291             -73.988597   14726  Subscriber   
7             40.751581             -73.977910   14691  Subscriber   
8             40.769155             -73.981918   14802  Subscriber   
9             40.689888             -73.981013   15018  Subscriber   

   birth_year  gender  
0        1973       1  
1        1950       1  
2        1983       1  
3        1987       2  
4        1970       1  
5   

In [7]:
# all unique start_station_id
task1_5 = '''
select distinct start_station_id
from citibike
order by start_station_id
'''
pd.read_csv(StringIO.StringIO(queryCartoDB(task1_5)), sep=',')

start_station_id  Unnamed: 1
0                  72         NaN
1                  79         NaN
2                  82         NaN
3                  83         NaN
4                 116         NaN
5                 119         NaN
6                 120         NaN
7                 127         NaN
8                 128         NaN
9                 137         NaN
10                143         NaN
11                144         NaN
12                146         NaN
13                147         NaN
14                150         NaN
15                151         NaN
16                152         NaN
17                153         NaN
18                157         NaN
19                160         NaN
20                161         NaN
21                164         NaN
22                167         NaN
23                168         NaN
24                173         NaN
25                174         NaN
26                195         NaN
27                212         NaN
28                216         NaN
29                217         NaN
..                ...         ...
297               525         NaN
298               526         NaN
299               527         NaN
300               528         NaN
301               529         NaN
302               530         NaN
303               531         NaN
304               532         NaN
305               533         NaN
306               534         NaN
307               537         NaN
308               538         NaN
309               539         NaN
310               545         NaN
311               546         NaN
312              2000         NaN
313              2001         NaN
314              2002         NaN
315              2004         NaN
316              2005         NaN
317              2006         NaN
318              2008         NaN
319              2009         NaN
320              2010         NaN
321              2012         NaN
322              2017         NaN
323              2021         NaN
324              2022         NaN
325              2023         NaN
326              3002         NaN

[327 rows x 2 columns]

In [8]:
# aggregation statistics data 
task1_6 = '''
select count(start_station_id) as trip_num,
avg(tripduration) as avg_duration,
max(tripduration) as max_duration,
min(tripduration) as min_duration
from citibike
'''
pd.read_csv(StringIO.StringIO(queryCartoDB(task1_6)), sep=',')

trip_num  avg_duration  max_duration  min_duration
0     46200    675.865823         43016            60

### Task 2 — Working with date/time
#### Selecting trips started on Feb-02-2015 only 
#### Selecting trips started on the weekends
    - What are average trip duration during weekends? 
#### Can we do the same for weekday?

In [9]:
# select trips started on 2-2-2015
task2_1 = '''
select * from citibike
where starttime < '2015-02-03 00:00'
and starttime >= '2015-02-02 00:00'
'''
pd.read_csv(StringIO.StringIO(queryCartoDB(task2_1)), sep=',')

the_geom           start_station_name  the_geom_webmercator  field_1  \
0          NaN              5 Ave & E 29 St                   NaN     7433   
1          NaN         Allen St & Hester St                   NaN     7779   
2          NaN            E 5 St & Avenue C                   NaN       11   
3          NaN              9 Ave & W 16 St                   NaN      627   
4          NaN              9 Ave & W 16 St                   NaN      991   
5          NaN              W 33 St & 7 Ave                   NaN     1037   
6          NaN              W 27 St & 7 Ave                   NaN     6510   
7          NaN              E 11 St & 1 Ave                   NaN     6511   
8          NaN         Peck Slip & Front St                   NaN     6512   
9          NaN              E 16 St & 5 Ave                   NaN     6513   
10         NaN              W 41 St & 8 Ave                   NaN     6514   
11         NaN              Dean St & 4 Ave                   NaN     6515   
12         NaN              9 Ave & W 18 St                   NaN     6516   
13         NaN          Reade St & Broadway                   NaN     6517   
14         NaN              8 Ave & W 31 St                   NaN     6518   
15         NaN              1 Ave & E 15 St                   NaN     6519   
16         NaN           Broadway & E 14 St                   NaN     6520   
17         NaN            E 2 St & Avenue B                   NaN     6521   
18         NaN         Peck Slip & Front St                   NaN     6522   
19         NaN   Columbia St & Rivington St                   NaN     6523   
20         NaN              W 20 St & 8 Ave                   NaN     6524   
21         NaN          St Marks Pl & 2 Ave                   NaN     6525   
22         NaN              W 20 St & 8 Ave                   NaN     6526   
23         NaN              W 37 St & 5 Ave                   NaN     6527   
24         NaN              W 13 St & 6 Ave                   NaN     6528   
25         NaN             10 Ave & W 28 St                   NaN     6529   
26         NaN              E 25 St & 2 Ave                   NaN     6530   
27         NaN             W 4 St & 7 Ave S                   NaN     6531   
28         NaN              W 20 St & 7 Ave                   NaN     6532   
29         NaN             W 46 St & 11 Ave                   NaN     6533   
...        ...                          ...                   ...      ...   
1429       NaN      Allen St & E Houston St                   NaN     7938   
1430       NaN           Broadway & W 29 St                   NaN     7939   
1431       NaN              W 56 St & 6 Ave                   NaN     7940   
1432       NaN             10 Ave & W 28 St                   NaN     7941   
1433       NaN              W 26 St & 8 Ave                   NaN     7942   
1434       NaN           Broadway & E 22 St                   NaN     7943   
1435       NaN              E 47 St & 2 Ave                   NaN     7944   
1436       NaN           Broadway & W 32 St                   NaN     7945   
1437       NaN              8 Ave & W 31 St                   NaN     7946   
1438       NaN              9 Ave & W 18 St                   NaN     7947   
1439       NaN        LaGuardia Pl & W 3 St                   NaN     7948   
1440       NaN              W 26 St & 8 Ave                   NaN     7949   
1441       NaN         Cliff St & Fulton St                   NaN     7950   
1442       NaN      Rivington St & Ridge St                   NaN     7951   
1443       NaN              W 49 St & 8 Ave                   NaN     7952   
1444       NaN        Pershing Square North                   NaN     7953   
1445       NaN         E 30 St & Park Ave S                   NaN     7954   
1446       NaN  Sullivan St & Washington Sq                   NaN     7955   
1447       NaN              W 42 St & 8 Ave                   NaN     7956   
1448       NaN           Bro

In [10]:
# trip on weekends
task2_2 = '''
select * from citibike
where extract(dow from starttime) in (0,6)
'''

pd.read_csv(StringIO.StringIO(queryCartoDB(task2_2)), sep=',')

the_geom        start_station_name  the_geom_webmercator  field_1  \
0           NaN        E 20 St & Park Ave                   NaN      175   
1           NaN          W 43 St & 10 Ave                   NaN     1159   
2           NaN         E 6 St & Avenue B                   NaN     2827   
3           NaN           E 12 St & 3 Ave                   NaN     4961   
4           NaN           W 41 St & 8 Ave                   NaN     6156   
5           NaN           E 20 St & 2 Ave                   NaN    39894   
6           NaN        Broadway & W 60 St                   NaN    41512   
7           NaN         E 6 St & Avenue D                   NaN    41595   
8           NaN            Great Jones St                   NaN    41806   
9           NaN           W 13 St & 6 Ave                   NaN    42316   
10          NaN          W 34 St & 11 Ave                   NaN    42532   
11          NaN           W 41 St & 8 Ave                   NaN    42915   
12          NaN           E 25 St & 1 Ave                   NaN    44448   
13          NaN           E 15 St & 3 Ave                   NaN    44744   
14          NaN           W 33 St & 7 Ave                   NaN    44834   
15          NaN           E 33 St & 2 Ave                   NaN    45301   
16          NaN           8 Ave & W 33 St                   NaN        0   
17          NaN        Broadway & W 39 St                   NaN        1   
18          NaN           W 18 St & 6 Ave                   NaN        2   
19          NaN        Broadway & E 22 St                   NaN        3   
20          NaN          W 46 St & 11 Ave                   NaN        4   
21          NaN            Great Jones St                   NaN        5   
22          NaN        Carmine St & 6 Ave                   NaN        6   
23          NaN           E 12 St & 3 Ave                   NaN        7   
24          NaN           W 20 St & 7 Ave                   NaN        8   
25          NaN           E 11 St & 1 Ave                   NaN        9   
26          NaN           W 24 St & 7 Ave                   NaN       10   
27          NaN           W 33 St & 7 Ave                   NaN       38   
28          NaN           W 17 St & 8 Ave                   NaN       49   
29          NaN           E 15 St & 3 Ave                   NaN       50   
...         ...                       ...                   ...      ...   
13201       NaN  Washington Pl & Broadway                   NaN    46170   
13202       NaN      Allen St & Hester St                   NaN    46171   
13203       NaN     Warren St & Church St                   NaN    46172   
13204       NaN           E 12 St & 3 Ave                   NaN    46173   
13205       NaN           Dean St & 4 Ave                   NaN    46174   
13206       NaN      Allen St & Hester St                   NaN    46175   
13207       NaN        Broadway & E 22 St                   NaN    46176   
13208       NaN     Washington Pl & 6 Ave                   NaN    46177   
13209       NaN           E 31 St & 3 Ave                   NaN    46178   
13210       NaN        Broadway & W 60 St                   NaN    46179   
13211       NaN        Carmine St & 6 Ave                   NaN    46180   
13212       NaN           W 52 St & 5 Ave                   NaN    46181   
13213       NaN           Dean St & 4 Ave                   NaN    46182   
13214       NaN           9 Ave & W 45 St                   NaN    46183   
13215       NaN        Broadway & W 58 St                   NaN    46184   
13216       NaN        Broadway & W 58 St                   NaN    46185   
13217       NaN   Watts St & Greenwich St                   NaN    46186   
13218       NaN           W 51 St & 6 Ave                   NaN    46187   
13219       NaN   University Pl & E 14 St                   NaN    46188   
13220       NaN           W 11 St & 6 Ave                   NaN    46189   
13221       NaN           W 17 St & 8 Ave              

In [11]:
# average duration
task2_3 = '''
select avg(tripduration) as avg_duration from citibike
where extract(dow from starttime) in (0,6)
'''
pd.read_csv(StringIO.StringIO(queryCartoDB(task2_3)), sep=',')

avg_duration  Unnamed: 1
0    662.942181         NaN

In [12]:
# weekdays 
task2_4 = '''
select * from citibike
where extract(dow from starttime) not in (0,6)
'''
pd.read_csv(StringIO.StringIO(queryCartoDB(task2_4)), sep=',')

the_geom              start_station_name  the_geom_webmercator  \
0           NaN                 5 Ave & E 29 St                   NaN   
1           NaN            Allen St & Hester St                   NaN   
2           NaN                 E 39 St & 3 Ave                   NaN   
3           NaN                10 Ave & W 28 St                   NaN   
4           NaN                  E 2 St & 2 Ave                   NaN   
5           NaN                 1 Ave & E 44 St                   NaN   
6           NaN                 W 45 St & 6 Ave                   NaN   
7           NaN                 W 18 St & 6 Ave                   NaN   
8           NaN                 E 58 St & 3 Ave                   NaN   
9           NaN                 E 33 St & 5 Ave                   NaN   
10          NaN                 W 31 St & 7 Ave                   NaN   
11          NaN                 5 Ave & E 29 St                   NaN   
12          NaN                 E 39 St & 2 Ave                   NaN   
13          NaN                10 Ave & W 28 St                   NaN   
14          NaN                 W 42 St & 8 Ave                   NaN   
15          NaN                 E 27 St & 1 Ave                   NaN   
16          NaN                  Great Jones St                   NaN   
17          NaN                 W 49 St & 8 Ave                   NaN   
18          NaN              Broadway & W 24 St                   NaN   
19          NaN                 E 40 St & 5 Ave                   NaN   
20          NaN                 E 16 St & 5 Ave                   NaN   
21          NaN              Broadway & W 36 St                   NaN   
22          NaN                 8 Ave & W 31 St                   NaN   
23          NaN                 W 45 St & 6 Ave                   NaN   
24          NaN             FDR Drive & E 35 St                   NaN   
25          NaN                 6 Ave & W 33 St                   NaN   
26          NaN                 E 27 St & 1 Ave                   NaN   
27          NaN                12 Ave & W 40 St                   NaN   
28          NaN                 W 52 St & 9 Ave                   NaN   
29          NaN                 8 Ave & W 33 St                   NaN   
...         ...                             ...                   ...   
32939       NaN          W Broadway & Spring St                   NaN   
32940       NaN              Broadway & W 41 St                   NaN   
32941       NaN                 W 38 St & 8 Ave                   NaN   
32942       NaN                 9 Ave & W 22 St                   NaN   
32943       NaN  Lafayette Ave & Fort Greene Pl                   NaN   
32944       NaN              Broadway & E 22 St                   NaN   
32945       NaN                 E 23 St & 1 Ave                   NaN   
32946       NaN                 E 19 St & 3 Ave                   NaN   
32947       NaN              Broadway & E 22 St                   NaN   
32948       NaN      Greenwich Ave & Charles St                   NaN   
32949       NaN            Allen St & Hester St                   NaN   
32950       NaN              E 20 St & Park Ave                   NaN   
32951       NaN         DeKalb Ave & Hudson Ave                   NaN   
32952       NaN                 E 16 St & 5 Ave                   NaN   
32953       NaN                 W 44 St & 5 Ave                   NaN   
32954       NaN                 E 56 St & 3 Ave                   NaN   
32955       NaN               E 2 St & Avenue B                   NaN   
32956       NaN                  E 4 St & 2 Ave                   NaN   
32957       NaN              E 17 St & Broadway                   NaN   
32958       NaN                 1 Ave & E 15 St                   NaN   
32959       NaN                 W 51 St & 6 Ave                   NaN   
32960       NaN                 W 11 St & 6 Ave                   NaN   
32961       NaN                 W 20 St & 8 Ave                  

In [13]:
# weekdays average

task2_5 = '''
select avg(tripduration) from citibike
where extract(dow from starttime) not in (0,6)

'''
pd.read_csv(StringIO.StringIO(queryCartoDB(task2_5)), sep=',')


avg  Unnamed: 1
0  681.052292         NaN

### Task 3 — Working with Space
#### Showing the list of start station locations Using GROUP BY
#### Showing the number of trips started per station
#### ... but only for stations within 500m of Time Square!
#### The coordinates of Time Square is (40.7577,-73.9857)

In [14]:
# a list of start locations
task3_1 = '''
select distinct on (start_station_id)
start_station_id,start_station_latitude,start_station_longitude 
from citibike
GROUP BY start_station_id, start_station_latitude, start_station_longitude
'''
pd.read_csv(StringIO.StringIO(queryCartoDB(task3_1)), sep=',')

start_station_id  start_station_latitude  start_station_longitude
0                  72               40.767272               -73.993929
1                  79               40.719116               -74.006667
2                  82               40.711174               -74.000165
3                  83               40.683826               -73.976323
4                 116               40.741776               -74.001497
5                 119               40.696089               -73.978034
6                 120               40.686768               -73.959282
7                 127               40.731724               -74.006744
8                 128               40.727103               -74.002971
9                 137               40.761628               -73.972924
10                143               40.692395               -73.993379
11                144               40.698399               -73.980689
12                146               40.716250               -74.009106
13                147               40.715422               -74.011220
14                150               40.720874               -73.980858
15                151               40.721816               -73.997203
16                152               40.714740               -74.009106
17                153               40.752062               -73.981632
18                157               40.690893               -73.996123
19                160               40.748238               -73.978311
20                161               40.729170               -73.998102
21                164               40.753231               -73.970325
22                167               40.748901               -73.976049
23                168               40.739713               -73.994564
24                173               40.760647               -73.984427
25                174               40.738177               -73.977387
26                195               40.709056               -74.010434
27                212               40.743349               -74.006818
28                216               40.700379               -73.995481
29                217               40.702772               -73.993836
..                ...                     ...                      ...
297               525               40.755942               -74.002116
298               526               40.747659               -73.984907
299               527               40.744023               -73.976056
300               528               40.742909               -73.977061
301               529               40.757570               -73.990985
302               530               40.771522               -73.990541
303               531               40.718939               -73.992663
304               532               40.710451               -73.960876
305               533               40.752996               -73.987216
306               534               40.702551               -74.012723
307               537               40.740259               -73.984092
308               538               40.756660               -73.976684
309               539               40.715348               -73.960241
310               545               40.736502               -73.978095
311               546               40.744449               -73.983035
312              2000               40.702551               -73.989402
313              2001               40.699773               -73.979927
314              2002               40.716887               -73.963198
315              2004               40.724399               -74.004704
316              2005               40.705312               -73.971001
317              2006               40.765909               -73.976342
318              2008               40.705693               -74.016777
319              2009               40.711174               -73.996826
320              2010               40.721655               -74.002347
321              2012        

In [19]:
# number of trips started per station within 500m of time square 
task3_2 = '''
select count(start_station_id),
start_station_name as start_station_name
from citibike
where st_dwithin(
CDB_LatLng(start_station_latitude, start_station_longitude)::geography, CDB_LatLng(40.7577, -73.9857)::geography, 500)
GROUP BY start_station_id, start_station_name
'''

pd.read_csv(StringIO.StringIO(queryCartoDB(task3_2)), sep=',')

count  start_station_name
0    213  Broadway & W 49 St
1    251  Broadway & W 41 St
2    507     W 41 St & 8 Ave
3    141     W 45 St & 6 Ave
4    112     W 43 St & 6 Ave
5    221     W 42 St & 8 Ave
6    141     W 45 St & 8 Ave

### Task 4 — Putting it all together
#### Find the station that had the longest average trip duration during weekends and within 500m of TimeSquare!
#### Extra: create lines for trips started from stations within 500m of Times Squares and lasted less than 2 hours. The number of trips per each pair of stations are output as attributes of these lines.

In [20]:
task4_1 = '''
select avg(tripduration) as avg_durations, start_station_name
from citibike
where st_dwithin(CDB_LatLng(start_station_latitude, start_station_longitude)::geography, CDB_LatLng(40.7577, -73.9857)::geography, 500) 
and extract(DOW FROM starttime) in (0,6)
group by start_station_latitude, start_station_longitude, start_station_name
order by avg_durations DESC
'''
pd.read_csv(StringIO.StringIO(queryCartoDB(task4_1)), sep=',')

avg_durations  start_station_name
0    1010.104167  Broadway & W 49 St
1     762.931818     W 45 St & 8 Ave
2     683.121212  Broadway & W 41 St
3     675.400000     W 45 St & 6 Ave
4     643.260274     W 41 St & 8 Ave
5     629.700000     W 43 St & 6 Ave
6     579.142857     W 42 St & 8 Ave

### Therefore, the start station with the longest average trip duration is Broadway & W 49 St with 1010.104167

In [20]:
# Extra Credit 
extra_credit = '''

select st_makeline(CDB_LatLng(start_station_latitude,start_station_longitude), CDB_LatLng(end_station_latitude,end_station_longitude)) as the_geom_webmercator,
count(start_station_id) as num_trips, start_station_name, end_station_name
from citibike
where st_dwithin(CDB_LatLng(start_station_latitude, start_station_longitude)::geography, CDB_LatLng(40.7577,-73.9857)::geography, 500) and tripduration < 60 * 60 * 2
group by start_station_latitude, start_station_longitude, end_station_latitude, end_station_longitude, start_station_name, end_station_name
'''
pd.read_csv(StringIO.StringIO(queryCartoDB(extra_credit)), sep=',')

the_geom_webmercator  num_trips  \
0    0102000020E610000002000000BE51C821247F52C02199...          1   
1    0102000020E610000002000000BE51C821247F52C02199...          1   
2    0102000020E610000002000000BE51C821247F52C02199...          1   
3    0102000020E610000002000000BE51C821247F52C02199...          1   
4    0102000020E610000002000000BE51C821247F52C02199...          1   
5    0102000020E610000002000000BE51C821247F52C02199...          1   
6    0102000020E610000002000000BE51C821247F52C02199...          3   
7    0102000020E610000002000000BE51C821247F52C02199...          1   
8    0102000020E610000002000000BE51C821247F52C02199...          1   
9    0102000020E610000002000000BE51C821247F52C02199...          1   
10   0102000020E610000002000000BE51C821247F52C02199...          1   
11   0102000020E610000002000000BE51C821247F52C02199...          4   
12   0102000020E610000002000000BE51C821247F52C02199...          3   
13   0102000020E610000002000000BE51C821247F52C02199...          1   
14   0102000020E610000002000000BE51C821247F52C02199...          1   
15   0102000020E610000002000000BE51C821247F52C02199...          1   
16   0102000020E610000002000000BE51C821247F52C02199...          2   
17   0102000020E610000002000000BE51C821247F52C02199...          2   
18   0102000020E610000002000000BE51C821247F52C02199...          1   
19   0102000020E610000002000000BE51C821247F52C02199...          1   
20   0102000020E610000002000000BE51C821247F52C02199...          1   
21   0102000020E610000002000000BE51C821247F52C02199...          2   
22   0102000020E610000002000000BE51C821247F52C02199...          2   
23   0102000020E610000002000000BE51C821247F52C02199...          2   
24   0102000020E610000002000000BE51C821247F52C02199...          1   
25   0102000020E610000002000000BE51C821247F52C02199...          1   
26   0102000020E610000002000000BE51C821247F52C02199...          6   
27   0102000020E610000002000000BE51C821247F52C02199...          2   
28   0102000020E610000002000000BE51C821247F52C02199...          2   
29   0102000020E610000002000000BE51C821247F52C02199...          5   
..                                                 ...        ...   
542  0102000020E610000002000000385762D8007F52C0BF6D...         13   
543  0102000020E610000002000000385762D8007F52C0BF6D...          1   
544  0102000020E610000002000000385762D8007F52C0BF6D...          2   
545  0102000020E610000002000000385762D8007F52C0BF6D...          5   
546  0102000020E610000002000000385762D8007F52C0BF6D...          1   
547  0102000020E610000002000000385762D8007F52C0BF6D...          1   
548  0102000020E610000002000000385762D8007F52C0BF6D...          1   
549  0102000020E610000002000000385762D8007F52C0BF6D...          2   
550  0102000020E610000002000000385762D8007F52C0BF6D...         12   
551  0102000020E610000002000000385762D8007F52C0BF6D...          6   
552  0102000020E610000002000000385762D8007F52C0BF6D...          9   
553  0102000020E610000002000000385762D8007F52C0BF6D...          6   
554  0102000020E610000002000000385762D8007F52C0BF6D...          2   
555  0102000020E610000002000000385762D8007F52C0BF6D...          1   
556  0102000020E610000002000000385762D8007F52C0BF6D...          2   
557  0102000020E610000002000000385762D8007F52C0BF6D...          2   
558  0102000020E610000002000000385762D8007F52C0BF6D...          2   
559  0102000020E610000002000000385762D8007F52C0BF6D...          4   
560  0102000020E610000002000000385762D8007F52C0BF6D...          1   
561  0102000020E610000002000000385762D8007F52C0BF6D...          2   
562  0102000020E610000002000000385762D8007F52C0BF6D...          1   
563  0102000020E610000002000000385762D8007F52C0BF6D...          2   
564  0102000020E610000002000000385762D8007F52C0BF6D...          1   
565  0102000020E610000002000000385762D8007F52C0BF6D...          1   
566  0102000020E610000002000000385762D8007F52C0BF6D...          1   
567  0102000020E610000002000000385762D8007F52C0BF6D...          3   
568  0102000020E6100000020000003

In [1]:
## Extra Credit: Don't have time to do the python2 and python3 things